File: ros_ws/src/cpp_pubsub/src/publisher_member_function.cpp
#include <chrono>
#include <functional>
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

using namespace std::chrono_literals;

class MinimalPublisher : public rclcpp::Node
{
	public:
		MinimalPublisher()
			: Node("minimal_publisher")
			, m_count { 0 }
		{
			constexpr size_t max_queue_length { 10 };
			m_publisher = this->create_publisher<std_msgs::msg::String>("topic", max_queue_length);
			m_timer = this->create_wall_timer(500ms, std::bind(&MinimalPublisher::timer_callback, this));
		}

	private:
		void timer_callback()
		{
			auto message = std_msgs::msg::String();
			message.data = "Hello, world!" + std::to_string(m_count);
			++m_count;
			RCLCPP_INFO(this->get_logger(), "Publishing: '%s'", message.data.c_str());
			m_publisher->publish(message);
		}

		rclcpp::TimerBase::SharedPtr m_timer;
		rclcpp::Publisher<std_msgs::msg::String>::SharedPtr m_publisher;
		size_t m_count;
};

int main(int argc, char* argv[])
{
	rclcpp::init(argc, argv);
	rclcpp::spin(std::make_shared<MinimalPublisher>());
	rclcpp::shutdown();
	return 0;
}

File: ros_ws/src/py_pubsub/py_pubsub/publisher_member_function.py
import rclpy
from rclpy.node import Node

from std_msgs.msg import String


class MinimalPublisher(Node):
    def __init__(self):
        super(MinimalPublisher, self).__init__('minimal_publisher')
        max_queue_length = 10
        self.publisher = self.create_publisher(
            String, 'topic', max_queue_length
        )
        timer_period_sec = 0.5
        self.timer = self.create_timer(timer_period_sec, self.timer_callback)
        self.count = 0

    def timer_callback(self):
        message = String()
        message.data = f'Hello World: {self.count}'
        self.count += 1
        self.publisher.publish(message)
        self.get_logger().info(f'Publishing: "{message.data}"')


def main(args=None):
    rclpy.init(args=args)

    minimal_publisher = MinimalPublisher()

    rclpy.spin(minimal_publisher)

    # Explicit for control (otherwise gc will destroy)
    minimal_publisher.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()
